src/detect-flow-pkts.c

```diff
@@ -23,238 +23,412 @@
+        return -1;
 
-static int PrefilterSetupFlowPktsToClient(DetectEngineCtx *de_ctx, SigGroupHead *sgh)
-{
-    return PrefilterSetupPacketHeader(de_ctx, sgh, DETECT_FLOW_PKTS_TO_CLIENT,
-            SIG_MASK_REQUIRE_FLOW, PrefilterPacketU32Set, PrefilterPacketU32Compare,
-            PrefilterPacketFlowPktsToClientMatch);
-}
+    df->pkt_data = du32;
+    df->dir = DETECT_FLOW_TOCLIENT;
 
-static bool PrefilterFlowPktsToClientIsPrefilterable(const Signature *s)
-{
-    return PrefilterIsPrefilterableById(s, DETECT_FLOW_PKTS_TO_CLIENT);
-}
+    if (SigMatchAppendSMToList(
+                de_ctx, s, DETECT_FLOW_PKTS, (SigMatchCtx *)df, DETECT_SM_LIST_MATCH) == NULL) {
+        DetectFlowPktsFree(de_ctx, df);
+        return -1;
+    }
+    s->flags |= SIG_FLAG_REQUIRE_PACKET;
 
-void DetectFlowPktsToClientRegister(void)
-{
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].name = "flow.pkts_toclient";
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].desc = "match flow number of packets to client";
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].url = "/rules/flow-keywords.html#flow-pkts_toclient";
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].Match = DetectFlowPktsToClientMatch;
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].Setup = DetectFlowPktsToClientSetup;
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].Free = DetectFlowPktsToClientFree;
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].SupportsPrefilter =
-            PrefilterFlowPktsToClientIsPrefilterable;
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].SetupPrefilter = PrefilterSetupFlowPktsToClient;
+    return 0;
 }
 
-static int DetectFlowPktsToServerMatch(
-        DetectEngineThreadCtx *det_ctx, Packet *p, const Signature *s, const SigMatchCtx *ctx)
+static int DetectFlowPktsSetup(DetectEngineCtx *de_ctx, Signature *s, const char *rawstr)
 {
-    if (p->flow == NULL) {
-        return 0;
+    DetectFlow *df = NULL;
+    char copy[strlen(rawstr) + 1];
+    strlcpy(copy, rawstr, sizeof(copy));
+    char *context = NULL;
+    char *token = strtok_r(copy, ",", &context);
+    uint8_t num_tokens = 1;
+    uint8_t dir = 0;
+    char *pkt_data = NULL;
+
+    while (token != NULL) {
+        if (num_tokens > 2)
+            goto error;
+
+        while (*token != '\0' && isblank(*token)) {
+            token++;
+        }
+        if (strlen(token) == 0) {
+            goto next;
+        }
+
+        num_tokens++;
+
+        if (strcmp(token, "toserver") == 0) {
+            dir = DETECT_FLOW_TOSERVER;
+        } else if (strcmp(token, "toclient") == 0) {
+            dir = DETECT_FLOW_TOCLIENT;
+        } else if (strcmp(token, "either") == 0) {
+            dir = DETECT_FLOW_TOEITHER;
+        }
+
+        if (dir) {
+            pkt_data = token;
+        }
+    next:
+        token = strtok_r(NULL, ",", &context);
     }
-    uint32_t nb = p->flow->todstpktcnt;
 
-    const DetectU32Data *du32 = (const DetectU32Data *)ctx;
-    return DetectU32Match(nb, du32);
+    if (dir) {
```

Comment:
What happens if `dir == 0` ?

Reply:
Made the default dir as toserver so removing this check altogether as a dir will always exist